src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs

```diff
@@ -56,149 +67,301 @@ public sealed class RichTextEditorPastedImages
 
-            var fileName = Path.GetFileName(absoluteTempImagePath);
-            var safeFileName = fileName.ToSafeFileName(_shortStringHelper);
-
-            var mediaItemName = safeFileName.ToFriendlyName();
-            IMedia mediaFile;
-            GuidUdi udi;
+            var ext = dataUriInfo.Groups["ext"].Value.ToLowerInvariant();
+            var encoding = dataUriInfo.Groups["encoding"].Value.ToLowerInvariant();
+            var imageData = dataUriInfo.Groups["data"].Value;
 
-            if (uploadedImages.ContainsKey(tmpImgPath) == false)
+            if (_contentSettings.IsFileAllowedForUpload(ext) == false)
             {
-                if (mediaParentFolder == Guid.Empty)
-                {
-                    mediaFile = _mediaService.CreateMedia(mediaItemName, Constants.System.Root, Constants.Conventions.MediaTypes.Image, userId);
-                }
-                else
-                {
-                    mediaFile = _mediaService.CreateMedia(mediaItemName, mediaParentFolder, Constants.Conventions.MediaTypes.Image, userId);
-                }
-
-                var fileInfo = new FileInfo(absoluteTempImagePath);
+                // If the image format is not supported we should probably leave it be
+                // since the user decided to include it.
+                // If we accepted it anyway, they could technically circumvent the allow list for file types,
+                // but the user experience would not be very good if we simply failed to save the content.
+                // Besides, there may be other types of data uri images technically supported by a browser that we cannot handle.
+                _logger.LogWarning(
+                    "Performance impact: Could not convert embedded image to a Media item because the file extension {Ext} was not allowed. HTML extract: {OuterHtml}",
+                    ext,
+                    img.OuterHtml.Length < 100 ? img.OuterHtml : img.OuterHtml[..100]); // only log the first 100 chars because base64 images can be very long
+                continue;
+            }
 
-                FileStream? fileStream = fileInfo.OpenReadWithRetry();
-                if (fileStream == null)
-                {
-                    throw new InvalidOperationException("Could not acquire file stream");
-                }
+            // Create an unique folder path to help with concurrent users to avoid filename clash
+            var imageTempPath =
+                _hostingEnvironment.MapPathContentRoot(Constants.SystemDirectories.TempImageUploads + Path.DirectorySeparatorChar + Guid.NewGuid());
 
-                using (fileStream)
-                {
-                    mediaFile.SetValue(_mediaFileManager, _mediaUrlGenerators, _shortStringHelper, _contentTypeBaseServiceProvider, Constants.Conventions.Media.File, safeFileName, fileStream);
-                }
+            // Ensure image temp path exists
+            if (Directory.Exists(imageTempPath) == false)
```

Comment:
Lets change this to pattern matching while we're at it 🔥
```suggestion
            if (Directory.Exists(imageTempPath) is false)
```